SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorPackageAssembly.cpp

```diff
@@ -0,0 +1,209 @@
+
+#include "CoreMinimal.h"
+
+#include "Async/Async.h"
+#include "Misc/App.h"
+#include "Misc/FileHelper.h"
+#include "UnrealEdMisc.h"
+#include "EditorStyle.h"
+#include "SpatialGDKSettings.h"
+#include "SpatialGDKEditorModule.h"
+#include "SpatialGDKEditorSettings.h"
+#include "SpatialGDKServicesModule.h"
+#include "Framework/Notifications/NotificationManager.h"
+#include "UATHelper/Public/IUATHelperModule.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialGDKEditorPackageAssembly);
+
+#define LOCTEXT_NAMESPACE "SpatialGDKEditorPackageAssembly"
+
+namespace
+{
+	const FString SpatialBuildExe = FSpatialGDKServicesModule::GetSpatialGDKPluginDirectory(TEXT("SpatialGDK/Binaries/ThirdParty/Improbable/Programs/Build.exe"));
+	const FString Linux = TEXT("Linux");
+}
+
+
+FSpatialGDKPackageAssembly::FSpatialGDKPackageAssembly() : CurrentAssemblyTarget{ EPackageAssemblyTarget::NONE }
+{
+
+}
+
+static FString GetStagingDir()
+{
+	return FPaths::ConvertRelativePathToFull((FPaths::IsProjectFilePathSet() ? FPaths::GetPath(FPaths::GetProjectFilePath()) : FPaths::RootDir() / FApp::GetProjectName()) / TEXT("..") / TEXT("spatial") / TEXT("build") / TEXT("unreal"));
+}
+
+void FSpatialGDKPackageAssembly::BuildAssembly(const FString& ProjectName, const FString& Platform, const FString& Configuration, const FString& AdditionalArgs)
+{
+	const USpatialGDKSettings* SpatialGDKSettings = GetDefault<USpatialGDKSettings>();
+	FString WorkingDir = FPaths::ConvertRelativePathToFull(FPaths::IsProjectFilePathSet() ? FPaths::GetPath(FPaths::GetProjectFilePath()) : FPaths::RootDir() / FApp::GetProjectName());
+	FString Project = FPaths::ConvertRelativePathToFull(FPaths::GetProjectFilePath());
+	FString Args = FString::Printf(TEXT("%s %s %s %s"), *ProjectName, *Platform, *Configuration, *Project, *AdditionalArgs);
+	PackageAssemblyTask = MakeShareable(new FMonitoredProcess(SpatialBuildExe, Args, WorkingDir, true));
+	PackageAssemblyTask->OnCompleted().BindRaw(this, &FSpatialGDKPackageAssembly::OnTaskCompleted);
+	PackageAssemblyTask->OnOutput().BindRaw(this, &FSpatialGDKPackageAssembly::OnTaskOutput);
+	PackageAssemblyTask->OnCanceled().BindRaw(this, &FSpatialGDKPackageAssembly::OnTaskCanceled);
+	PackageAssemblyTask->Launch();
+	FString NotificationMessage = FString::Printf(TEXT("Building %s Assembly"), *ProjectName);
+}
+
+void FSpatialGDKPackageAssembly::UploadAssembly(const FString &AssemblyName, bool Force)
+{
+	const USpatialGDKSettings* SpatialGDKSettings = GetDefault<USpatialGDKSettings>();
+	FString WorkingDir = FPaths::ConvertRelativePathToFull((FPaths::IsProjectFilePathSet() ? FPaths::GetPath(FPaths::GetProjectFilePath()) : FPaths::RootDir() / FApp::GetProjectName()) / TEXT("..") / TEXT("spatial"));
+	FString Spatial(TEXT("spatial"));
+	FString Args = FString::Printf(TEXT("cloud upload %s %s %s --no_animation"), *AssemblyName, Force ? TEXT("--force") : TEXT(""), SpatialGDKSettings->IsRunningInChina() ? TEXT("--environment=cn-production") : TEXT(""));
```

Comment:
No, because the whole purpose of this class is to construct an FMonitoredProcess and allow it's cancellation, all the notification work and state management is here, putting this call into the SpatialCommandUtils would not follow the pattern of commands there, and we would have to add communication via delegates which I believe is unnecessary. This is not a reusable command that other people should be calling.